### DISCUSSION TITLE
LinkedIn provider should expose entire profile

### DISCUSSION THREAD

**Discussion by ghost** [nextauthjs/next-auth#2689] [Help]
### Description 📓

Currently the LinkedIn provider doesn't expose all the attributes returned by LinkedIn.

### How to reproduce ☕️

Standard example with LinkedIn provider.

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by balazsorban44** [comment]
https://next-auth.js.org/providers/linkedin#options

**Comment by ghost** [comment]
@balazsorban44 Why was this closed?

**Comment by balazsorban44** [comment]
Have you checked the link? Have a particular look at the `profile` callback.

**Reply by ghost** [reply]
I had a look at the link but still a little confused on what the best way to surface firstName, lastName and emailAddress separately.

**Reply by balazsorban44** [reply]
I updated our docs, please have a look: https://next-auth.js.org/configuration/providers/oauth-provider#override-default-options

**Reply by balazsorban44** [reply]
the `session` and `jwt` callbacks might also be of interest: https://next-auth.js.org/configuration/callbacks

**Reply by ghost** [reply]
Thanks for the update. I figured out that I need to override the authorization URL to include the r_emailaddress scope. I am still not sure how to return the firstName and lastName separately.

**Reply by balazsorban44** [reply]
Could you paste in some code you tried?

**Reply by ghost** [reply]
```js
import NextAuth from "next-auth";
import LinkedInProvider from "next-auth/providers/linkedin";
import axios from "axios";

export default NextAuth({
	// Configure one or more authentication providers
	providers: [
		LinkedInProvider({
			clientId: process.env.LINKEDIN_CLIENT_ID,
			clientSecret: process.env.LINKEDIN_CLIENT_SECRET,
			authorization:
				"https://www.linkedin.com/oauth/v2/authorization?scope=r_liteprofile%20r_emailaddress",
			profile: async (profile, tokens) => {
				const res = await fetch(
					`https://api.linkedin.com/v2/emailAddress?q=members&projection=(elements*(handle~))&oauth2_access_token=${tokens.access_token}`
				);
				const data = await res.json();
				const email = data.elements[0]["handle~"].emailAddress;
				return {
					id: profile.id,
					firstName: profile.localizedFirstName,
					lastName: profile.localizedLastName,
					name: profile.localizedFirstName + " " + profile.localizedLastName,
					email,
					image: null,
				};
			},
		}),
	],
	options: {
		debug: true,
	},
});

```

**Reply by balazsorban44** [reply]
I cannot verify that your profile callback returns values correctly, but assuming you double-checked it, it looks correct!

I am guessing you want these values in the front-end. Have you checked the `session` and `jwt` callbacks yet? See link above

**Reply by ghost** [reply]
I get the email address - that part of the profile callback works. The firstName and lastName fields get dropped even though I am adding them in the profile callback.

I looked at the docs for `session` and `jwt` but wasn't sure how they effect the firstName and lastName. I will take another look.

**Reply by balazsorban44** [reply]
Where do you stop seeing the `firstName` and `lastName`? Is it already gone in the `profile` callback? Or your problem is not seeing them in `useSession`? If it's the former, check your scopes. If the latter, then the answer will be the callbacks.
